docs(x/accounts/defaults/lockup): Add tutorial document (#22168)
cosmos/cosmos-sdk: x/accounts/defaults/lockup/TUTORIAL.md

* [Setup](#setup)
* [Init](#init)
* [Execution](#execution)
    * [Delegate](#delegate)
    * [Undelegate](#undelegate)
    * [Withdraw reward](#withdraw-reward)
    * [Withdraw unlocked token](#withdraw-unlocked-token)
    * [Send coins](#send-coins)
* [Query](#query)
    * [Query account info](#query-account-info)
    * [Query periodic lockup account locking periods](#query-periodic-lockup-account-locking-periods)

To learn more about lockup account, please also check out [readme](./README.md)

## Setup 

To create a lockup account we need 2 wallets (newly created or use any of the existing wallet that you have) one for the creator and one for the owner of the lockup account. 

```bash
simd keys add creator 
simd keys add owner
```

## Init

Normally the creator must have enough token to grant to the lockup account during the lockup account init process. The owner wallet should be associated with the individual that the creator want to grant the fund to.

Now, the creator can craft the lockup account init messages. This message depend on what type of lockup account the creator want to create.
For continous, delayed, permanent locking account:

> ```go

Reviewer: ```suggestion
```json
```

{
    "owner": "cosmos1vaqh39cdex9sgr69ef0tdln5cn0hdyd3s0lx45",
    "end_time": 1495793860
    "start_time": 1465793854
}
```

:::info
`start_time` is only needed for continous locking account init process. For the other two, you dont have to set it in. Error will returned if `start_time` is not provided when creating continous locking account*
:::
 
For periodic locking account:

> ```go

Reviewer: ```suggestion
```json
```

To initialize the account, we have to run the accounts init command passing the account type and the json string for the init message.

```bash
initcontents=$(cat init.json)
simd tx accounts init <lockup_type> $initcontents --from creator
```

Whereas the available `lockup_type` options are: 

* continuous-locking-account

* delayed-locking-account

* periodic-locking-account

* permanent-locking-account

If success, we'll check the tx result for the lockup account address. You can send token to it like a normal account. 

## Execution

To execute a message, we can use the command below:

```bash
msgcontents=$(cat msg.json)
simd tx accounts execute <account_address> <execute-msg-type-url> $msgcontents --from owner
```

Whereas `execute-msg-type-url` and `msgcontents`  corresponds to lockup account available executions, which are:

### Delegate

The execute message type url for this execution is `cosmos.accounts.defaults.lockup.MsgDelegate`.

Example of json file:

> ```go

Reviewer: ```suggestion
```json
```

{
    "sender": "cosmos1vaqh39cdex9sgr69ef0tdln5cn0hdyd3s0lx45",
    "validator_address": "cosmosvaloper1vaqh39cdex9sgr69ef0tdln5cn0hdyd3s0lx45",
    "amount": {
        "amount": 100
        "denom": "stake"
    }
}
``` 

:::warning
The `sender` field are the address of the owner of the lockup account. If the sender is not the owner an error will be returned.
:::

### Undelegate

The execute message type url for this execution is `cosmos.accounts.defaults.lockup.MsgUndelegate`.

Example of json file:

> ```go

Reviewer: ```suggestion
```json
```

{
    "sender": "cosmos1vaqh39cdex9sgr69ef0tdln5cn0hdyd3s0lx45",
    "validator_address": "cosmosvaloper1vaqh39cdex9sgr69ef0tdln5cn0hdyd3s0lx45",
    "amount": {
        "amount": 100
        "denom": "stake"
    }
}
``` 

:::warning
The `sender` field are the address of the owner of the lockup account. If the sender is not the owner an error will be returned.
:::

### Withdraw reward 

The execute message type url for this execution is `cosmos.accounts.defaults.lockup.MsgWithdrawReward`.

Example of json file:

> ```go

Reviewer: ```suggestion
```json
```

{
    "sender": "cosmos1vaqh39cdex9sgr69ef0tdln5cn0hdyd3s0lx45",
    "validator_address": "cosmosvaloper1vaqh39cdex9sgr69ef0tdln5cn0hdyd3s0lx45",
}
```

:::warning
The `sender` field are the address of the owner of the lockup account. If the sender is not the owner an error will be returned.
:::

### Withdraw unlocked token

The execute message type url for this execution is `cosmos.accounts.defaults.lockup.MsgWithdraw`.

Example of json file:

> ```go

Reviewer: ```suggestion
```json
```

{
    // lockup account owner address
    "withdrawer": "cosmos1vaqh39cdex9sgr69ef0tdln5cn0hdyd3s0lx46",
    // withdraw to an account of choice
    "to_address": "cosmos1vaqh39cdex9sgr69ef0tdln5cn0hdyd3s0lx47",
    "denoms": ["stake"]
}
``` 

:::warning
The `withdrawer` field are the address of the owner of the lockup account. If the sender is not the owner an error will be returned.
:::

### Send coins

The execute message type url for this execution is `cosmos.accounts.defaults.lockup.MsgSend`.

Example of json file:

> ```go

Reviewer: ```suggestion
```json
```

# Using lockup account on Cosmos sdk

* [Setup](#setup)
* [Init](#init)
* [Execution](#execution)
    * [Delegate](#delegate)
    * [Undelegate](#undelegate)
    * [Withdraw reward](#withdraw-reward)
    * [Withdraw unlocked token](#withdraw-unlocked-token)
    * [Send coins](#send-coins)
* [Query](#query)
    * [Query account info](#query-account-info)
    * [Query periodic lockup account locking periods](#query-periodic-lockup-account-locking-periods)

> 

Reviewer 2: Let's add an abstract/summary to link the readme and have a quick refresher on what this is.

## Setup 

To create a lockup account we need 2 wallets (newly created or use any of the existing wallet that you have) one for the granter and one for the owner of the lockup account. 

```bash

> simd keys add granter --keyring-backend test --home ./.testnets/node0/simd/

Reviewer 3: can we reword from granter to admin or something like that? granter sounds weird and will confuse people

Reviewer 3: maybe creator would be better

Reviewer 2: +1 and let's drop all the `--home` and `--keyring-backend` flags. Those are redundant, and the not necessary for the tutorial

simd keys add owner --keyring-backend test --home ./.testnets/node0/simd/
```

## Init

> Normally the granter must have enough token to grant to the lockup account during the lockup account init process. The owner wallet should be associated with the individual that the granter want to grant the fund to.

Reviewer 3: ```suggestion
Normally the granter must have enough token to create to the lockup account during the lockup account init process. The owner wallet should be associated with the individual that the granter wants to create the lockup account for.
```

Now, the granter can craft the lockup account init messages. This message depend on what type of lockup account the granter want to create.

> </br>

Reviewer 3: ```suggestion
```

Reviewer 2: Nope, should be removed

For continous, delayed, permanent locking account:

```go
{
    "owner": "cosmos1vaqh39cdex9sgr69ef0tdln5cn0hdyd3s0lx45",
    "end_time": 1495793860
    "start_time": 1465793854
}
```

> :::infor

Reviewer 2: ```suggestion
:::info
```

> *note: `start_time` is only needed for continous locking account init process. For the other two, you dont have to set it in. Error will returned if `start_time` is not provided when creating continous locking account*

Reviewer 2: ```suggestion
`start_time` is only needed for continous locking account init process. For the other two, you dont have to set it in. Error will returned if `start_time` is not provided when creating continous locking account*
```

:::

> </br>

Reviewer 2: to delete
